controller/volume_controller.go

```diff
@@ -859,18 +859,34 @@ func (c *VolumeController) ReconcileEngineReplicaState(v *longhorn.Volume, es ma
 			continue
 		}
 
-		if status.State == engineapi.ProcessStateComplete && v.Status.CloneStatus.State != longhorn.VolumeCloneStateCompleted {
-			v.Status.CloneStatus.State = longhorn.VolumeCloneStateCompleted
-			c.eventRecorder.Eventf(v, corev1.EventTypeNormal, constant.EventReasonVolumeCloneCompleted,
-				"finished cloning snapshot %v from source volume %v",
+		if !isCloneTargetCopyInProgress(v) {
+			// No longer need to sync up with the engine
+			continue
+		}
+
+		switch status.State {
+		case engineapi.ProcessStateComplete:
+			v.Status.CloneStatus.State = longhorn.VolumeCloneStateCopyCompletedAwaitingHealthy
+			c.eventRecorder.Eventf(
+				v, corev1.EventTypeWarning, constant.EventReasonVolumeCloneCopyCompleteAwaitingHealthy,
+				"copied the data from snapshot %v of the source volume %v. Waiting for volume to be fully HA before marking the clone as completed",
 				v.Status.CloneStatus.Snapshot, v.Status.CloneStatus.SourceVolume)
-		} else if status.State == engineapi.ProcessStateError && v.Status.CloneStatus.State != longhorn.VolumeCloneStateFailed {
+		case engineapi.ProcessStateError:
 			v.Status.CloneStatus.State = longhorn.VolumeCloneStateFailed
-			c.eventRecorder.Eventf(v, corev1.EventTypeWarning, constant.EventReasonVolumeCloneFailed,
+			c.eventRecorder.Eventf(
+				v, corev1.EventTypeWarning, constant.EventReasonVolumeCloneFailed,
 				"failed to clone snapshot %v from source volume %v: %v",
-				v.Status.CloneStatus.Snapshot, v.Status.CloneStatus.SourceVolume, status.Error)
+				v.Status.CloneStatus.Snapshot, v.Status.CloneStatus.SourceVolume, status.Error,
+			)
 		}
```

Comment:
Since we’re looping over `e.Status.CloneStatus` but only updating `v.Status.CloneStatus.State`, if there are multiple entries, won’t the last one just overwrite the earlier updates and cause weird state transitions?

Reply:
No because there would be max 1 replica when doing cloning